### PR TITLE
refactor(backend): migrate components and projects services from Thrift to Spring Boot REST

### PR DESCRIPTION
**Pull request by Shivamrut**
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

# Change Documentation — Components & Projects Service Migration

## Summary

Migrate the **components** and **projects** backend services from Thrift RPC to Spring Boot REST **together**, following the established pattern from health, configurations, search, changelogs, vendors, vmcomponents, SPDX, packages, attachments, fossology, vulnerabilities, users, and licenses migrations.

This PR replaces the Thrift servlets for both services with Spring Boot WARs at `/components/api/components` and `/projects/api/projects`, introduces Thrift-`Iface`-compatible REST adapters in resource-server, moves `ComponentHandler` / `ProjectHandler` into `backend-common` so backend callers can use in-process access, and removes the Thrift servlets / `web.xml` / `index.jsp` for both modules.

Unlike the users migration (which also migrated the public HAL layer to service-api POJOs), this PR keeps the public `/resource/api/components`, `/resource/api/releases`, and `/resource/api/projects` HAL controllers on Thrift types. Conversion happens only at the backend REST boundary and inside the resource-server adapters. **No intentional public wire-contract change.**

Components and projects are migrated in one PR because they are tightly coupled (shared release graph, linking, export, moderation triggers) and migrating one without the other would leave a fragile half-state.

---

### Service-API POJOs (reused, not introduced in this PR)

Package roots under `org.eclipse.sw360.datahandler.services`:

| Package | Key types used |
|---|---|
| `components` | `Component`, `Release`, `ReleaseLink`, `ReleaseNode`, `BulkOperationNode`, … |
| `projects` | `Project`, `ClearingRequest`, `ProjectLink`, `ObligationList`, `UsedReleaseRelations`, … |
| `attachments` | `Attachment` |
| `common` | `PaginationData`, `PaginatedResult`, `RequestStatus`, `RequestSummary`, `AddDocumentRequestSummary`, `ProjectReleaseRelationship`, `ReleaseRelationship`, … |
| `users` | `RequestedAction` (and user identity carried via `X-User-*` headers) |

### Converters (backend/common, pre-existing)

Packages under `org.eclipse.sw360.common.utils.converter`:

| Package | Purpose |
|---|---|
| `components.*` | `ComponentConverter`, `ReleaseConverter`, `ReleaseLinkConverter`, `ReleaseNodeConverter`, `BulkOperationNodeConverter`, … |
| `projects.*` | `ProjectConverter`, `ClearingRequestConverter`, `ProjectLinkConverter`, `ObligationListConverter`, … |
| `attachments.*` | `AttachmentConverter` |
| `common.*` | `PaginationDataConverter`, `RequestStatusConverter`, `RequestSummaryConverter`, … |

Lossless Thrift ↔ POJO conversion at the REST boundary only. DB handlers remain Thrift-typed.

---

## Layer 2 — Backend Components Service (`backend/components`)

| File | Action | Purpose |
|---|---|---|
| `backend/components/pom.xml` | modified | Added `service-api`, `spring-boot-starter-web`, `spring-boot-starter-tomcat` (provided), `spring-boot-maven-plugin` for WAR packaging |
| `ComponentsApplication.java` | new | Spring Boot entry point; extends `SpringBootServletInitializer` for Tomcat WAR deploy |
| `ComponentController.java` | new | `@RestController` at `/api/components` — full REST surface replacing `ComponentService.Iface` (~108 endpoints, including releases nested under components) |
| `ComponentRestMapper.java` | new | POJO ↔ Thrift mapping for controller I/O; delegates to `backend-common` converters |
| `ComponentRestExceptionHandler.java` | new | `@ControllerAdvice` mapping `SW360Exception` error codes → HTTP status |
| `ComponentServlet.java` | deleted | Thrift servlet wrapper |
| `src/main/webapp/WEB-INF/web.xml` | deleted | Thrift servlet registration |
| `src/main/webapp/index.jsp` | deleted | Placeholder JSP |
| `application.yml` | new | `spring.application.name: components` (+ EPL-2.0 header) |
| `ComponentHandler.java` | **moved** | Relocated to `backend-common` (see Layer 4) |

Auth identity for privileged operations is passed via `X-User-Email` / `X-User-Department` / `X-User-Group` and rebuilt with `UserUtils.buildUser(...)`.

`ComponentDatabaseHandler` / CouchDB remain Thrift-based internally. Conversion is only at the REST controller boundary via `ComponentRestMapper`.

---

## Layer 3 — Backend Projects Service (`backend/projects`)

| File | Action | Purpose |
|---|---|---|
| `backend/projects/pom.xml` | modified | Same Spring Boot WAR packaging as components |
| `ProjectsApplication.java` | new | Spring Boot entry point |
| `ProjectController.java` | new | `@RestController` at `/api/projects` — full REST surface replacing `ProjectService.Iface` (~75 endpoints) |
| `ProjectRestMapper.java` | new | POJO ↔ Thrift mapping; paginated results returned as `PaginatedResult<Project>` |
| `ProjectRestExceptionHandler.java` | new | `@ControllerAdvice` for `SW360Exception` |
| `ProjectServlet.java` | deleted | Thrift servlet |
| `src/main/webapp/WEB-INF/web.xml` | deleted | Thrift servlet registration |
| `src/main/webapp/index.jsp` | deleted | Placeholder JSP |
| `application.yml` | new | `spring.application.name: projects` (+ EPL-2.0 header) |
| `ProjectHandler.java` | **moved** | Relocated to `backend-common` (constructors made public) |

Same `X-User-*` header pattern as components.

---

## Layer 4 — Handler relocation (`backend-common`)

| File | Action | Purpose |
|---|---|---|
| `backend/common/.../components/ComponentHandler.java` | moved (from `backend/components`) | Shared JAR so backend WARs can call components in-process |
| `backend/common/.../projects/ProjectHandler.java` | moved (from `backend/projects`) | Shared JAR so backend WARs can call projects in-process |
| `ComponentDatabaseHandler.java` / `ProjectDatabaseHandler.java` | modified | Call-site updates for relocated handlers / in-process access |
| `CycloneDxBOMExporter.java` | modified | In-process handler access |

**Why move handlers into `backend-common`?**  
`backend/components` and `backend/projects` are WAR modules. Other backend modules cannot cleanly depend on WARs as compile-time JARs. Putting the handlers in `backend-common` (already a shared JAR) enables in-process calls without circular WAR dependencies — same practical goal as the users `UsersClients` library, but for co-located backend callers that already share the DB layer.

---

## Layer 5 — Resource-Server adapters (Thrift Iface shim)

| File | Action | Purpose |
|---|---|---|
| `ComponentServiceRestAdapter.java` | new | Spring `@Component` implementing `ComponentService.Iface`; HTTP to `/components/api/components`; Thrift ↔ POJO conversion |
| `ProjectServiceRestAdapter.java` | new | Spring `@Component` implementing `ProjectService.Iface`; HTTP to `/projects/api/projects`; Thrift ↔ POJO conversion |
| `Sw360ComponentService.java` | modified | Inject `ComponentServiceRestAdapter` instead of `ThriftClients.makeComponentClient()` |
| `Sw360ProjectService.java` | modified | Inject `ProjectServiceRestAdapter` / `ComponentServiceRestAdapter` |
| `Sw360ReleaseService.java` | modified | Adapter injection |
| `Sw360AttachmentService.java` | modified | Adapter injection |
| `Sw360AttachmentCleanUpService.java` | modified | Adapter injection |
| `Sw360DatabaseSanitationService.java` | modified | Adapter injection |
| `Sw360ImportExportService.java` | modified | Adapter injection |
| `Sw360ModerationRequestService.java` | modified | Adapter injection |
| `SW360ReportService.java` | modified | Adapter injection |
| `Sw360VendorService.java` | modified | Adapter injection |
| `Sw360VulnerabilityService.java` | modified | Adapter injection |

**Adapter contract notes:**
- Adapters preserve the Thrift `Iface` so existing resource-server services need minimal signature changes.
- Paginated endpoints use `POST` + JSON body (`PaginationData` / filter DTOs) to match the backend controllers (not GET+query-params).
- Public HAL controllers (`ComponentController`, `ProjectController`, `ReleaseController` under `resource-server`) stay on Thrift types — wire JSON unchanged.

---

## Layer 6 — Backend cross-service callers (in-process)

| File | Action | Purpose |
|---|---|---|
| `backend/fossology/.../FossologyHandler.java` | modified | `new ComponentHandler()` instead of Thrift client |
| `backend/fossology/.../FossologyHandlerTest.java` | modified | `throws TException` on setup after handler signature change |
| `backend/licenseinfo/.../LicenseInfoHandler.java` | modified | In-process component/project handler access |
| `backend/licenseinfo/.../DocxGenerator.java` | modified | In-process access |
| `backend/schedule/.../ScheduleHandler.java` | modified | `new ComponentHandler()` / `new ProjectHandler()` for scheduled jobs |
| `backend/vendors/.../VendorDatabaseHandler.java` | modified | In-process `ComponentHandler` |
| `backend/wsimport/.../ThriftExchange.java` | modified | In-process handlers for component/project upload paths |

---

## Layer 7 — Deprecation (`libraries/datahandler`)

| File | Action | Purpose |
|---|---|---|
| `ThriftClients.java` | modified | `@Deprecated` on `makeComponentClient()` / `makeProjectClient()` with Javadoc pointing to REST adapters / in-process handlers |
| `SW360Utils.java` | modified | `@Deprecated` on helpers that still call `makeComponentClient()` / `makeProjectClient()` (`getLinkedProjects`, `getLinkedReleases`, `getUsingProjectByReleaseIds`, …) |

These deprecated `SW360Utils` helpers remain temporarily because they live in the lowest library layer and still use Thrift clients internally. Callers that already sit in `backend-common` / resource-server were redirected; leftover low-layer helpers are marked for later cleanup.

---

## Layer 8 — Tests

| File | Action | Purpose |
|---|---|---|
| `Sw360AttachmentServiceTest.java` | modified | Adapter / injection wiring |
| `Sw360ProjectServiceTest.java` | modified | Adapter / injection wiring |
| `SW360ReportServiceTest.java` | modified | Adapter / injection wiring |
| `FossologyHandlerTest.java` | modified | Compile fix for handler signature |

Existing component / project / release integration + restdocs suites continue to mock resource-server services; they do not exercise every backend internal REST path end-to-end.

---

### Components — `http://localhost:8080/components/api/components`

Representative groups (full surface ≈ Thrift `ComponentService` method set, ~108 mappings):

| Group | Examples |
|---|---|
| Component CRUD | `GET /{id}`, `POST /`, `PUT /`, `PUT /force`, `DELETE /{id}`, `DELETE /{id}/force` |
| Accessibility / edit | `GET /accessible/{id}`, `GET /{id}/for-edit`, `GET /accessible/{id}/for-edit` |
| Search / pagination | `POST /search`, `POST /search/accessible/paginated`, `POST /recent/paginated`, `POST /search-by-external-ids` |
| Subscriptions | `GET /my`, `GET /subscribed`, `POST /{id}/subscribe`, `POST /{id}/unsubscribe` |
| Releases (nested) | release CRUD / search / linking endpoints under the same controller |
| Export / reports | `GET /download-excel`, `GET /report-stream`, `GET /report-email`, summary/export paths |
| Merge / moderation | `POST /merge`, `PUT /moderation`, `PUT /bulk` |

User-scoped operations require `X-User-Email` (+ optional department/group).

### Projects — `http://localhost:8080/projects/api/projects`

Representative groups (~75 mappings):

| Group | Examples |
|---|---|
| Project CRUD | `GET /{id}`, `POST /`, `PUT /`, `PUT /force`, `DELETE /{id}`, `DELETE /{id}/force` |
| Search / filters | `GET /search`, `POST /search/refined`, `POST /search/refined/paginated`, by-name / by-group / by-tag / by-type |
| Linking | `POST /linked`, `GET /{id}/linking`, by-release / by-package paths |
| Clearing / obligations | clearing-request and obligation-list related endpoints |
| Counts / groups | `GET /my-count`, `GET /groups`, count-by-release / package paths |
| Moderation | `PUT /moderation` |

---

## Public HAL API (`/resource/api/...`)

**Unchanged paths and wire contract** (resource-server HAL controllers still serialize Thrift types):

| Area | Paths (unchanged) |
|---|---|
| Components | `/resource/api/components`, `/resource/api/{id}`, nested releases, search, create/update/delete, … |
| Releases | `/resource/api/releases`, … |
| Projects | `/resource/api/projects`, nested releases/packages/attachments/licenseInfo, … |

**What changed internally (not wire contract):**
- `Sw360ComponentService` / `Sw360ProjectService` / related services talk to backends via `ComponentServiceRestAdapter` / `ProjectServiceRestAdapter` instead of Thrift
- Backend WARs speak POJO JSON on `/components/api/...` and `/projects/api/...`
- Thrift servlets `/components/thrift` and `/projects/thrift` are removed (expect **HTTP 404** after clean deploy)

---

### 1. Components + projects migrated together (intentional big-bang within the incremental program)

Mutual coupling (release linking, project↔release maps, export, CycloneDX, schedule jobs, vendor updates) makes a one-service-first approach riskier. This PR is a controlled dual migration so both Thrift servlets disappear in the same change set.

### 2. Thrift `Iface` adapters in resource-server (unlike users HAL POJO migration)

Users migrated the public HAL layer to service-api POJOs because auth/security already needed POJO `User`. Components/projects still dominate the HAL surface with deeply nested Thrift graphs (`Component` ↔ `Release` ↔ attachments ↔ vendors). Keeping HAL on Thrift and converting only in adapters minimizes public-API risk for this transport step.

### 3. In-process handlers for backend callers (vs dedicated `*-rest-client` library)

Users needed a shared client library because callers span resource-server, authorization-server, rest-common, and many backends. For components/projects, backend callers already share `backend-common` and the DB layer, so `new ComponentHandler()` / `new ProjectHandler()` is the lowest-overhead replacement for Thrift RPC between backend WARs. Resource-server still uses HTTP adapters (different process boundary).

### 4. DB handlers stay on Thrift internally

Per project decision for transport migrations: no POJO migration inside `ComponentDatabaseHandler` / `ProjectDatabaseHandler`. Conversion is only at the REST controller / adapter boundary.

### 5. Paginated endpoints as POST + JSON body

Thrift paginated methods take structured `PaginationData` (+ filters). Mapping them to GET+query-params was error-prone; controllers use POST with JSON bodies, and adapters match that contract exactly.

### 6. Deprecate, don’t delete, `ThriftClients` / `SW360Utils` helpers yet

`makeComponentClient()` / `makeProjectClient()` are deprecated. Remaining `SW360Utils` helpers that still call them are also deprecated for a follow-up cleanup once lowest-layer callers are reworked.

### 7. No behavioral change intended

Public HAL paths, status codes, and JSON shapes should remain compatible. Internal transport is REST; domain model in DB + HAL remains Thrift until a later core-removal phase.

---

## How Everything Connects

```text
Browser / Client
    |
    v
resource-server (resource.war)
    |  HAL: /resource/api/components|releases|projects
    |       (Thrift types, unchanged wire)
    |  Sw360ComponentService / Sw360ProjectService / Sw360ReleaseService / ...
    |       +-- ComponentServiceRestAdapter / ProjectServiceRestAdapter
    |            (implements ComponentService.Iface / ProjectService.Iface)
    |            HTTP + POJO JSON -> /components/api/... and /projects/api/...
    v
backend/components (components.war)          backend/projects (projects.war)
    |  ComponentController                     |  ProjectController
    |  ComponentRestMapper (POJO <-> Thrift)   |  ProjectRestMapper (POJO <-> Thrift)
    |  ComponentHandler (from backend-common)  |  ProjectHandler (from backend-common)
    |  ComponentDatabaseHandler                |  ProjectDatabaseHandler
    v                                          v
CouchDB (sw360db)

Parallel in-process path (no HTTP):
backend/fossology, schedule, vendors, licenseinfo, wsimport, cyclonedx, ...
    |  new ComponentHandler() / new ProjectHandler()
    v
backend-common handlers -> DatabaseHandlers -> CouchDB
```
---

### What I verified locally

| Check | Result |
|---|---|
| Full build + deploy (`mvn clean install -P deploy -DskipTests`) | Success |
| Tomcat clean redeploy | Success |
| Spot checks: by-id, nested releases, name search, pagination | HTTP 200 |
| Write round-trip (create → get → delete → get) on a throwaway component | 201 / 200 / 207 / 404 |
| Legacy Thrift endpoints `/components/thrift`, `/projects/thrift` | HTTP **404** (expected — removed) |
| Other still-Thrift backends (e.g. `/moderation/thrift`, `/licenseinfo/thrift`) | Still deployed (HTTP 500 on empty POST probe) |

### Testing limitation — review help requested

This migration exposes **100+ internal backend REST endpoints** (components ≈ 108 + projects ≈ 75), plus the existing large public HAL surface for components, releases, and projects.

I could **not** exhaustively test and verify every endpoint / edge case individually. I validated build/deploy, health checks, representative read/write/search/pagination paths, and Thrift-servlet removal.

**Request to reviewers:** please help exercise functional coverage for this commit — especially:
- Component / release CRUD, merge, bulk update, subscribe/unsubscribe
- Project linking, clearing-related flows, package/release associations
- Import/export, reports, CycloneDX / Fossology / schedule paths that call components/projects
- Permission / accessibility variants (`accessible/*`, force delete/update)
- Paginated and refined search endpoints

### CODE REVIEW

File: backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java [hunk 1]
```diff
@@ -376,31 +376,27 @@ public RequestStatus updateProject(Project project, User user) throws TException
      * Throws TException if any empty key is found.
      */
     private void validateNoEmptyKeys(Project project) throws TException {
-        // Additional Role (roles)
```
Comment (GMishx) [inline backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java:379]: Please do not remove comments.

File: backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java [hunk 2]
```diff
@@ -376,31 +376,27 @@ public RequestStatus updateProject(Project project, User user) throws TException
...
         if (project.isSetRoles() && project.getRoles() != null) {
             for (Map.Entry<String, java.util.Set<String>> entry : project.getRoles().entrySet()) {
                 if (entry.getKey() == null || entry.getKey().trim().isEmpty()) {
                     throw new TException("Project roles contain empty key");
                 }
             }
         }
-        // External Urls
```
Comment (GMishx) [inline backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java:387]: Please do not remove documentation comments.

File: backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java [hunk 3]
```diff
@@ -376,31 +376,27 @@ public RequestStatus updateProject(Project project, User user) throws TException
...
         if (project.isSetExternalUrls() && project.getExternalUrls() != null) {
             for (Map.Entry<String, String> entry : project.getExternalUrls().entrySet()) {
                 if (entry.getKey() == null || entry.getKey().trim().isEmpty()) {
                     throw new TException("External Urls contain empty key");
                 }
             }
         }
-        // External Ids
```
Comment (GMishx) [inline backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java:395]: Please do not remove documentation comments.

File: backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java [hunk 4]
```diff
@@ -376,31 +376,27 @@ public RequestStatus updateProject(Project project, User user) throws TException
...
         if (project.isSetExternalIds() && project.getExternalIds() != null) {
             for (Map.Entry<String, String> entry : project.getExternalIds().entrySet()) {
                 if (entry.getKey() == null || entry.getKey().trim().isEmpty()) {
                     throw new TException("External Ids contain empty key");
                 }
             }
         }
-        // Additional Data
```
Comment (GMishx) [inline backend/common/src/main/java/org/eclipse/sw360/projects/ProjectHandler.java:403]: Please do not remove documentation comments.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by Shivamrut** [comment]
Adding sequence number for order of migration: service-19 and service-20

**Comment by Shivamrut** [comment]
Hi @GMishx Restored the comments. Rebased the PR with latest merge. Please review. Also I will need assistance testing this PRs functionality (refer pr description)

**Review by GMishx — APPROVED** [review]
LGTM
